### PR TITLE
Support Dynamic Partition Overwrite for Indirect write

### PR DESCRIPTION
**Pull request by isha97**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -861,6 +861,17 @@ word-break:break-word
      </td>
      <td>Read</td>
    </tr>
+  <tr>
```
Comment (davidrabinowitz) [inline README.md:864]: This should be in the README-template only, not README

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 2]
```diff
@@ -648,10 +648,11 @@ public void loadDataIntoTable(
       List<String> sourceUris,
       FormatOptions formatOptions,
       JobInfo.WriteDisposition writeDisposition,
-      Optional<Schema> schema) {
+      Optional<Schema> schema,
+      TableId tableToWrite) {
```
Comment (davidrabinowitz) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:652]: nit: rename to `destinationTable` or similar

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 3]
```diff
@@ -105,7 +124,23 @@ public void writeDataFrameToBigQuery() {
       String format = config.getIntermediateFormat().getDataSource();
       data.write().format(format).save(gcsPath.toString());
 
-      loadDataToBigQuery();
+      if (writeDisposition == JobInfo.WriteDisposition.WRITE_TRUNCATE
+          && config.getPartitionOverwriteModeValue() == PartitionOverwriteMode.DYNAMIC
+          && bigQueryClient.tableExists(config.getTableId())) {
+        temporaryTableId = createTempTable(config.getTableId(), tableSchema);
+        loadDataToBigQuery();
+        Job queryJob =
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:132]: `overwriteDestinationWithTemporaryDynamicPartitons()` should be of type void, run the job and wait for it.

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 4]
```diff
@@ -105,7 +124,23 @@ public void writeDataFrameToBigQuery() {
...
+            bigQueryClient.overwriteDestinationWithTemporaryDynamicPartitons(
+                temporaryTableId, config.getTableId());
+        BigQueryClient.waitForJob(queryJob);
+        Preconditions.checkState(
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:136]: The table deletion should be initiated in the `BigQueryClient.createTempTable()` method, you can add the snippet 
```
      // Registering a cleanup job
      CLEANUP_JOBS.add(() -> deleteTable(tempTable));
```

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 5]
```diff
@@ -63,6 +67,11 @@ public class BigQueryWriteHelper {
   private final Path gcsPath;
   private final Optional<IntermediateDataCleaner> createTemporaryPathDeleter;
 
+  private final Schema tableSchema;
+  private final JobInfo.WriteDisposition writeDisposition;
+
+  private TableId temporaryTableId = null;
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:73]: Should be `Optional<TableId> temporaryTableId = Optional.empty()`

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 6]
```diff
@@ -105,7 +124,23 @@ public void writeDataFrameToBigQuery() {
       String format = config.getIntermediateFormat().getDataSource();
       data.write().format(format).save(gcsPath.toString());
 
-      loadDataToBigQuery();
+      if (writeDisposition == JobInfo.WriteDisposition.WRITE_TRUNCATE
+          && config.getPartitionOverwriteModeValue() == PartitionOverwriteMode.DYNAMIC
+          && bigQueryClient.tableExists(config.getTableId())) {
+        temporaryTableId = createTempTable(config.getTableId(), tableSchema);
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:130]: `temporaryTableId = Optional.of(createTempTable(config.getTableId(), tableSchema));`

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 7]
```diff
@@ -128,16 +178,18 @@ void loadDataToBigQuery() throws IOException {
     List<String> optimizedSourceUris = SparkBigQueryUtil.optimizeLoadUriListForSpark(sourceUris);
     JobInfo.WriteDisposition writeDisposition =
         SparkBigQueryUtil.saveModeToWriteDisposition(saveMode);
-    Schema schema =
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:131]: Why was this removed?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java [hunk 8]
```diff
@@ -128,16 +178,18 @@ void loadDataToBigQuery() throws IOException {
...
-        SchemaConverters.from(SchemaConvertersConfiguration.from(config))
-            .toBigQuerySchema(data.schema());
-    if (tableInfo.isPresent()) {
-      schema =
-          BigQueryUtil.adjustSchemaIfNeeded(schema, tableInfo.get().getDefinition().getSchema());
+    if (temporaryTableId != null) {
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:181]: Simpler:
```
TableId destinationTableId = temporaryTableId.orElse(config.getTableId())
bigQueryClient.loadDataIntoTable(
          config,
          optimizedSourceUris,
          formatOptions,
          writeDisposition,
          Optional.of(tableSchema),
          destinationTableId);
```

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java [hunk 9]
```diff
@@ -93,18 +96,36 @@ abstract class WriteIntegrationTestBase extends SparkBigQueryIntegrationTestBase
...
+  protected boolean sparkVersionAboveOrEqualTo3_4;
+
+  protected DataType timeStampNTZType;
 
   public WriteIntegrationTestBase(SparkBigQueryConfig.WriteMethod writeMethod) {
-    this(writeMethod, IllegalArgumentException.class);
+    this(writeMethod, IllegalArgumentException.class, false, null);
+  }
+
+  public WriteIntegrationTestBase(
```
Comment (davidrabinowitz) [inline spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java:107]: No need for the flag, see Spark34ReadIntegrationTest for suggestion

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (isha97) [reply README.md]: Fixed

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java (hunk 2 above)
Comment (isha97) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java]: Done

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java (hunk 4 above)
Comment (isha97) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java]: Done

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java (hunk 7 above)
Comment (isha97) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java:131]: This was moved to the writeDataFrameToBigQuery() function to use the schema to create temp table

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java (hunk 8 above)
Comment (isha97) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/write/BigQueryWriteHelper.java]: Done

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java (hunk 9 above)
Comment (isha97) [reply spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java:105]: Fixed

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 10]
```diff
@@ -760,6 +772,16 @@ public void loadDataIntoTable(
     }
   }
 
+  public void loadDataIntoTable(
```
Comment (davidrabinowitz) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:775]: Please check if this method is used. It was reference twice in the code, and both use the longer signature

**Review by davidrabinowitz — APPROVED** [review]
One thing to fix and we can merge.
